python/lsst/cp/pipe/ptc.py

```diff
@@ -432,85 +485,141 @@ def funcAstier(self, pars, x):
         a00, gain, noise = pars
         return 0.5/(a00*gain*gain)*(np.exp(2*a00*x*gain)-1) + noise/(gain*gain)
 
-    def fitPtcAndNl(self, fitVectorsDict, ptcFitType='POLYNOMIAL'):
-        """Function to fit PTC, and calculate linearity and linearity residual
+    @staticmethod
+    def _initialParsForPolynomial(order):
+        assert(order >= 2)
+        pars = np.zeros(order, dtype=np.float)
+        pars[0] = 10
+        pars[1] = 1
+        pars[2:] = 0.0001
+        return pars
+
+    @staticmethod
+    def _boundsForPolynomial(initialPars):
+        lowers = [np.NINF for p in initialPars]
+        uppers = [np.inf for p in initialPars]
+        lowers[1] = 0  # no negative gains
+        return (lowers, uppers)
+
+    @staticmethod
+    def _boundsForAstier(initialPars):
+        lowers = [np.NINF for p in initialPars]
+        uppers = [np.inf for p in initialPars]
+        return (lowers, uppers)
+
+    @staticmethod
+    def _getInitialGoodPoints(means, variances, maxDeviation):
+        """Return a boolean array to mask bad points.
+
+        A linear function has a constant ratio, so find the median
+        value of the ratios, and exclude the points that deviate
+        from that by more than a factor of maxDeviation.
+
+        Too high and points that are so bad that fit will fail will be included
+        Too low and the non-linear points will be excluded, biasing the NL fit."""
+        ratios = [b/a for (a, b) in zip(means, variances)]
+        medianRatio = np.median(ratios)
+        ratioDeviations = [r/medianRatio for r in ratios]
+        goodPoints = [abs(1-r) < maxDeviation for r in ratioDeviations]
+        return np.array(goodPoints)
+
+    def fitPtcAndNl(self, dataset, ptcFitType):
+        """Fit the photon transfer curve and calculate linearity and residuals.
+
+        Fit the photon transfer curve with either a polynomial of order
+        specified in the task config, or using the Astier approximation.
```

Comment:
"...the Astier approximation (Eq. 16 of 1905.08677)".  Their paper has a full model (Eq. 20) which has an approximation (Eq. 15). This one, in turn, has another approximation (Eq. 16), which is the one we are using for now.

Reply:
I was just improving the docstring that you originally wrote here! 😄 I think that info is already in the main class docstring though, and it's clear here that we're talking about the same function. Maybe that one could be improved too though, with this information. If you think so, could you suggest exactly how you'd like the main class docstring to read, and I'll update that instead?

Reply:
OK, since it is already in the main class docstring, then it is fine. The paper and the equation are also already cited there.